### PR TITLE
pageserver: add disk consistent and remote lsn metrics

### PR DESCRIPTION
**Pull request by VladLazar**
## Problem

There's no metrics for disk consistent LSN and remote LSN. This stuff is useful when looking at ingest performance.

## Summary of changes

Two per timeline metrics are added: `pageserver_disk_consistent_lsn` and `pageserver_projected_remote_consistent_lsn`. I went for the projected remote lsn instead of the visible one
because that more closely matches remote storage write tput. Ideally we would have both, but these metrics are expensive.

### CODE REVIEW

File: pageserver/src/metrics.rs [hunk 1]
```diff
@@ -2475,7 +2494,11 @@ impl TimelineMetrics {
...
             &timeline_id,
         );
-        let last_record_gauge = LAST_RECORD_LSN
+        let last_record_lsn_gauge = LAST_RECORD_LSN
+            .get_metric_with_label_values(&[&tenant_id, &shard_id, &timeline_id])
+            .unwrap();
+
+        let disk_consistent_lsn_gauge = DISK_CONSISTENT_LSN
             .get_metric_with_label_values(&[&tenant_id, &shard_id, &timeline_id])
             .unwrap();
```
Comment (problame) [inline pageserver/src/metrics.rs:2503]: Need to add cleanup step, grep for `LAST_RECORD_LSN` to see how it's done.

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -3850,6 +3850,10 @@ impl Timeline {
     fn set_disk_consistent_lsn(&self, new_value: Lsn) -> bool {
         let old_value = self.disk_consistent_lsn.fetch_max(new_value);
         assert!(new_value >= old_value, "disk_consistent_lsn must be growing monotonously at runtime; current {old_value}, offered {new_value}");
+
+        self.metrics
+            .disk_consistent_lsn_gauge
+            .set(new_value.0 as i64);
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:3856]: NB that disk_consistent_lsn is kinda meaningless in practice. Not sure if we want to spend a time series per tenant on this.

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 3]
```diff
@@ -2190,6 +2190,14 @@ impl RemoteTimelineClient {
...
 
                     let lsn = upload_queue.clean.0.metadata.disk_consistent_lsn();
+                    crate::metrics::PROJECTED_REMOTE_CONSISTENT_LSN
+                        .get_metric_with_label_values(&[
+                            &self.tenant_shard_id.tenant_id.to_string(),
+                            &format!("{}", self.tenant_shard_id.shard_slug()),
+                            &self.timeline_id.to_string(),
+                        ])
+                        .unwrap()
+                        .set(lsn.0 as i64);
```
Comment (problame) [inline pageserver/src/tenant/remote_timeline_client.rs:2200]: Please integrate into `self.metrics`. Otherwise we won't clean up.

**Review by problame — CHANGES_REQUESTED** [review]
per-timeline metrics need to be cleaned up. Actually, some tests should be failing with the current state of the PR because there's not enough cleanup

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 3 above)
Comment (VladLazar) [reply pageserver/src/tenant/remote_timeline_client.rs:2200]: The metric itself is de-registered in `TimelineMetrics::shutdown`. The metrics shut down last and by that point the remote timeline client should have exited. I can thread the gauge through to here if you insist though.

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:3856]: Why? It tells us how much we are lagging by. If it's more than 10GiB. Compute will backpressure.
I want this in order to prove that the remote segment uploads are stalling L0 generation. I expect to see the remote projected lsn be equal to the disk consistent lsn.

File: pageserver/src/metrics.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/metrics.rs:2503]: Weird, somehow I failed to Cmd-F it.

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (problame) [reply pageserver/src/tenant/timeline.rs:3856]: >  It tells us how much we are lagging by.

What matters is the remote consistent LSN.

> I expect to see the remote projected lsn be equal to the disk consistent lsn.

Yeah. Ok, I don't mind. I think the VictoriaMetrics has enough capacity.

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 3 above)
Comment (problame) [reply pageserver/src/tenant/remote_timeline_client.rs:2200]: Yeah, I would prefer we don't get advanterous with the metrics lifecycles.

This should be part of `RemoteTimelineClientMetrics` and be hooked into its lifecycle machinery (`impl Drop for RemoteTimelineClientMetrics`).

Comment (problame) [reply pageserver/src/tenant/remote_timeline_client.rs:2200]: Don't forget to remove it from

https://github.com/neondatabase/neon/blob/5b376379e5c571ce686b8358670915f3396eff35/pageserver/src/metrics.rs#L2670-L2674

**Review by problame — APPROVED** [review]
Approving in expectation that https://github.com/neondatabase/neon/pull/10005#discussion_r1871653169 will be addressed

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 3 above)
Comment (VladLazar) [reply pageserver/src/tenant/remote_timeline_client.rs]: https://github.com/neondatabase/neon/pull/10005/commits/4f74ea9a9bd5c050ca37471dedf01376a57dc0b4

**Pull request merged**